README.md
# NB8080 - Notes Board 8080

[![Join the chat at https://gitter.im/nb8080/Lobby](https://badges.gitter.im/nb8080/Lobby.svg)](https://gitter.im/nb8080/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

A light notes board application similar to 
[Notepad8080](https://daftspaniel.github.io/demos/np8080/) in philosophy.

Clientside only - no server/syncing/spying etc!

TODO Feature list:
- Multiple notes
- Offline
- Colour picker
- Differ note types
- Download / Export view

You can [try it out here](https://daftspaniel.github.io/demos/nb8080/). 


CHANGELOG.md
# Changelog

## v0.0.1

- Getting started.
- Better drop positioning of notes.

web/pwa.dart
import 'package:pwa/worker.dart';
import 'package:nb8080/pwa/offline_urls.g.dart' as offline;

/// The Progressive Web Application's entry point.
void main() {
  // The Worker handles the low-level code for initialization, fetch API
  // routing and (later) messaging.
  Worker worker = new Worker();

  // The static assets that need to be in the cache for offline mode.
  // By default it uses the automatically generated list from the output of
  // `pub build`. To refresh this list, run `pub run pwa` after each new build.
  worker.offlineUrls = offline.offlineUrls;

  // The above list can be extended with additional URLs:
  //
  // List<String> offlineUrls = new List.from(offline.offlineUrls);
  // offlineUrls.addAll(['https://www.example.org/custom/resource/']);
  // worker.offlineUrls = offlineUrls;

  // Fine-tune the caching and network fetch with dynamic caches and cache
  // strategies on the url-prefixed network routes:
  //
  // DynamicCache cache = new DynamicCache('images');
  // worker.router.registerGetUrl('https://cdn.example.com/', cache.networkFirst);

  // Start the worker.
  worker.run(version: offline.lastModified);
}


web/lib/note.dart
import 'dart:convert';
import 'dart:html';

import 'board.dart';
import 'localstorageutil.dart';

class Note {
  Map plainNote = new Map();

  String id = "1";
  int x = 0;
  int y = 0;

  DivElement note;
  Board board;

  void set backColor(String value) {
    note.style.backgroundColor = value;
  }

  void set foreColor(String value) {
    note.style.color = value;
  }

  Note(DivElement newNote, String newId, Board newBoard) {
    id = newId;
    note = newNote;
    board = newBoard;

    note.onKeyUp.listen((KeyboardEvent k) {
      save();
    });

    note.onMouseDown.listen((MouseEvent e) {
      x = e.client.x - (note.offsetLeft);
      y = e.client.y - (note.offsetTop);
      note.style.transition = 'none';
      board.setActiveNote(this);
      board.putNoteColorsInPicker();
    });

    note.onDragStart.listen((e) {
      note.style.opacity = "0.2";
    });

    note.onDragEnd.listen((e) {
      note.style.opacity = "1";
      note.style.transition = 'top 0.5s, left 0.5s';
    });
  }

  void load() {
    String stored = getStoredValue(id, null);

    if (stored == null) {
      create();
    } else {
      plainNote = JSON.decode(stored);
      note.innerHtml = plainNote['text'];
      note.style.top = plainNote['top'];
      note.style.left = plainNote['left'];

      if (plainNote['color'] == null)
        plainNote['color'] = '#000000';
      if (plainNote['background-color'] == null)
        plainNote['background-color'] = '#ffffff';

      note.style.color = plainNote['color'];
      note.style.backgroundColor = plainNote['background-color'];
    }
  }

  void delete() {
    deleteStoredValue(id);
  }

  void create() {
    plainNote['text'] = "Welcome to Notes Board 8080";
    plainNote['top'] = "100px";
    plainNote['left'] = "100px";
    plainNote['color'] = board.selectedForeColor;
    plainNote['background-color'] = board.selectedBackColor;
    note.innerHtml = plainNote['text'];

    move(75, 75);
  }

  void save() {
    plainNote['text'] = note.innerHtml;
    plainNote['left'] = note.style.left;
    plainNote['top'] = note.style.top;
    plainNote['color'] = note.style.color;
    plainNote['background-color'] = note.style.backgroundColor;
    storeValue(id, JSON.encode(plainNote));
  }

  void move(int mx, int my) {
    note.style.top = "${my}px";
    note.style.left = "${mx}px";
    save();
  }

}

web/lib/board.dart
import 'dart:convert';
import 'dart:html';

import 'htmlutil.dart';
import 'localstorageutil.dart';
import 'note.dart';

class Board {

  final DivElement board = querySelector('#board');
  final InputElement backColor = querySelector('#backColorPick');
  final InputElement foreColor = querySelector('#foreColorPick');
  final List<Note> notes = new List<Note>();

  final int noteWidth = 200;
  int newId = -1;
  Note activeNote;
  List<int> Ids = new List<int>();

  Board() {
    setupEventHandlers();
    initState();
  }

  void initState() {
    newId = int.parse(getStoredValue("newId", "1"));
    backColor.value = getStoredValue('noteBackColor', '#f1f555');
    foreColor.value = getStoredValue('foreBackColor', '#000000');
  }

  void loadNotes() {
    String idJSON = getStoredValue('AllNoteIds', '');
    if (idJSON.length > 0) {
      Ids = JSON.decode(idJSON);
    }
    Ids.forEach((int id) {
      if (id > 0) addNote(id);
    });
  }

  void setupEventHandlers() {
    board.onDragOver.listen((MouseEvent e) {
      e.preventDefault();
    });

    board.onDrop.listen((MouseEvent e) {
      activeNote.move(e.page.x - activeNote.x, e.page.y - activeNote.y);
    });

    backColor.onChange.listen((Event e) {
      storeValue('noteBackColor', backColor.value);
      activeNote.backColor = backColor.value;
      activeNote.save();
    });

    foreColor.onChange.listen((Event e) {
      storeValue('foreBackColor', foreColor.value);
      activeNote.foreColor = foreColor.value;
      activeNote.save();
    });
  }

  void putNoteColorsInPicker() {
    backColor.value = getHexColor(activeNote.note.style.backgroundColor);
    foreColor.value = getHexColor(activeNote.note.style.color);
  }

  String get selectedBackColor {
    return backColor.value;
  }

  String get selectedForeColor {
    return foreColor.value;
  }

  String getNewNoteID() {
    int claimedId = newId;
    newId ++;
    storeValue('newId', newId.toString());
    Ids.add(claimedId);
    storeValue('AllNoteIds', JSON.encode(Ids));
    return claimedId.toString();
  }

  void addNote([int i = -1]) {
    DivElement newNoteDiv = new DivElement();
    newNoteDiv
      ..classes.add('note')
      ..draggable = true
      ..contentEditable = 'true'
      ..style.backgroundColor = backColor.value
      ..style.color = foreColor.value;
    board.append(newNoteDiv);
    newNoteDiv.focus();

    String id;
    Note newNote;
    if (i < 0) {
      id = getNewNoteID();
      newNote = new Note(newNoteDiv, id, this);
      newNote
        ..move(75, 75)
        ..save();
    }
    else {
      id = i.toString();
      newNote = new Note(newNoteDiv, id, this);
      newNote.load();
    }

    notes.add(newNote);
    setActiveNote(newNote);
    newNote.note.focus();
  }

  void setActiveNote(Note note) {
    activeNote = note;
    notes.forEach((Note note) => note.note.style.zIndex = '10');
    activeNote.note.style.zIndex = '100';
  }

  void removeActiveNote() {
    if (window.confirm(
        "Are you sure you want to delete this note?")) {
      if (activeNote != null) {
        Ids.remove(int.parse(activeNote.id));
        storeValue('AllNoteIds', JSON.encode(Ids));
        activeNote.delete();
        activeNote.note.remove();
        notes.remove(activeNote);
      }

      if (notes.length > 0) {
        activeNote = notes[0];
      }
    }
  }

  void arrangeNotes() {
    int x = 60;
    int y = 30;
    int sx = 60;
    int z = 0;
    notes.forEach((Note note) {
      note.move(x, y);
      x = x + noteWidth + 30;
      if (x + noteWidth > board.clientWidth) {
        y += 100 + 30;
        sx += 10;
        x = sx;
      }
      note.note.focus();
      note.note.style.zIndex = z.toString();
      z++;
    });
  }
}

